### PR TITLE
Incremental streaming and reasoning display

### PR DESCRIPTION
**Pull request by riesbri**
Two changes that turned out to be one state machine, plus a bug the real-terminal run exposed.

## Incremental streaming

The live region held the whole reply and re-escaped, re-split, and retransmitted all of it on every delta. Measured against the built renderer with 5-character deltas:

| reply | before | after |
| --- | --- | --- |
| 1 KB | 112 KB / 37 ms | 25 KB / 10 ms |
| 6 KB | 646 KB / 121 ms | 126 KB / 28 ms |
| 26 KB | 2703 KB / 519 ms | 512 KB / 87 ms |

`escapeControls` ran over the entire accumulated answer once per token, so the CPU term grew faster than the reply. TTY writes on Linux are synchronous, so it backpressured the event loop and delayed key handling and the agent loop.

A line now enters scrollback the moment its newline arrives and is never revisited; only the unfinished trailing line stays live. Cost is linear in the reply instead of quadratic, and a reply longer than the window scrolls the terminal normally rather than being clipped to a fixed 8-line tail.

Two caps keep the live region shorter than the screen: a character cut, so a reply that never emits a newline cannot grow the per-draw cost, and a row cut, because a full-width character fills two columns and would otherwise double the rows the character cut allows. Both are mutation-checked.

## Reasoning display

`reasoning-delta` chunks were dropped at the `text-delta` gate, so a reasoning model showed a spinner over an empty region for as long as it thought. They now render dimmed and italic above the answer, and are left unparsed — a half-formed thought is not a document. The first `text` delta is what marks reasoning finished; the log carries no separate marker.

## Turn-end reasons

The cancel branch tested `reason.kind === 'canceled'`. `TurnEndReasonMap` has no such member — the tag is `aborted` — so it had never fired, and a ctrl-c that visibly stopped a reply left nothing explaining why. `max-tokens` and `blocked` were silent for the same reason; a reply cut off at the output ceiling looked exactly like a finished one.

## Verification

- 182 unit tests, including 7 new frame tests through `@xterm/headless` asserting a streamed reply reaches scrollback exactly once no matter how the provider chunks it, and that a reply taller than the window keeps every line in the terminal's own scroll buffer.
- Six mutations applied to `stream.ts` and one to the harness, each confirmed to fail a test that names the behaviour.
- Real pseudo-terminal runs against the live API: reasoning visible mid-turn, a 60-line reply streaming with the chrome pinned below it, and an interrupt keeping the partial reply plus an `interrupted` note.

Also fixes the frame-test harness, which read rows from buffer index zero rather than the viewport's `baseY` — buffer-absolute, so once output scrolled it returned the oldest rows. Any test whose output fit on screen looked correct either way.

### CODE REVIEW

File: packages/tui/src/stream.ts [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+   * behind it.
+   * @param columns - the terminal's current width.
+   * @returns rows for the live region.
+   */
+  live(columns: number): string[] {
+    const channel = this.current
+    if (channel === undefined) return []
+    const state = this.channels[channel]
+    if (state.pending === '') return []
+    const budget = Math.max(MIN_LIVE_COLUMNS, columns - CONTINUATION.length)
```
[the remark this answers is not in the document]
Comment (riesbri) [reply packages/tui/src/stream.ts]: Confirmed and fixed. `Math.max(MIN_LIVE_COLUMNS, columns - 2)` returned 8 on a 5-column terminal, so every row came back wider than the screen and got wrapped again — four nominal rows becoming however many the overflow demanded, which is exactly the failure the row cap exists to prevent. The budget is now what is left after the gutter and never more, and a terminal too narrow for one content column gets no stream region rather than a broken one. Guarded by a test asserting no returned row exceeds the terminal width at 3, 4, 8, 12, 20 and 41 columns.

File: packages/tui/src/stream.ts [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+    const rows = wrapToWidth(channel === 'reasoning' ? style(escaped, 'dim', 'italic') : escaped, budget)
+    const shown = rows.slice(-LIVE_ROWS)
+    const elided = shown.length < rows.length || visible.length < state.pending.length
+    const [first = '', ...rest] = shown
+    const head = state.opened ? CONTINUATION : `${style(MARK[channel], channel === 'reasoning' ? 'gray' : 'green')} `
+    return [
+      // The blank spacer belongs to the mark: once the mark is committed, the
+      // live rows continue lines directly above them and must stay attached.
+      ...state.opened ? [] : [''],
+      `${head}${elided ? style(ELLIPSIS, 'gray') : ''}${first}`,
```
[the remark this answers is not in the document]
Comment (riesbri) [reply packages/tui/src/stream.ts]: Confirmed and fixed. The first shown row is already exactly as wide as the budget allows, so the marker now comes out of that row's content via `truncateToWidth(first, budget - 1)` rather than being added beside it. Covered by the same width assertion.

File: packages/tui/src/stream.ts [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+    if (this.current !== undefined && this.current !== channel) out.push(...this.flush(this.current))
+    this.current = channel
+    const state = this.channels[channel]
+    state.pushed += delta
+    state.pending += delta
+    const cut = state.pending.lastIndexOf('\n')
+    if (cut < 0) return out
+    const complete = state.pending.slice(0, cut)
+    state.pending = state.pending.slice(cut + 1)
+    out.push(...this.emit(channel, complete.split('\n')))
```
[the remark this answers is not in the document]
Comment (riesbri) [reply packages/tui/src/stream.ts:151]: Confirmed and fixed, and this was the most valuable of the three — it was a genuine asymmetry between the streamed and assembled paths that my own symmetry test missed, because its fixture had no leading or trailing blank. Blank rows are now held until a non-blank row flushes them, dropped before the mark exists, and discarded if the turn ends still holding them. One refinement worth noting: blankness is judged on the RENDERED row rather than the source line, because inside a fenced block a blank source line renders as indented code and is content, while in prose it renders to nothing. There is now a test that chunks `\n\nHello\n\n\nWorld\n\n` at four different sizes and asserts every one equals the unchunked render.

File: packages/tui/src/stream.ts [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+    // for even one content column gets no stream region rather than a broken one.
+    const budget = columns - CONTINUATION.length
+    if (budget < 1) return []
+    // Cut the source before wrapping. A reply that never emits a newline would
+    // otherwise make each redraw cost the whole reply again, which is the
+    // quadratic term this class exists to remove; two columns per character is
+    // the widest any character gets, so this keeps every row that can be shown.
+    const visible = state.pending.slice(-budget * LIVE_ROWS)
+    const escaped = escapeControls(visible)
+    const rows = wrapToWidth(channel === 'reasoning' ? style(escaped, 'dim', 'italic') : escaped, budget)
```
[the remark this answers is not in the document]
Comment (riesbri) [reply packages/tui/src/stream.ts:232]: Right again, and thank you. `wrapToWidth` has to emit a two-column glyph at a one-column budget — refusing would make no progress and never terminate — so the wrap cannot be trusted to respect the budget it was given.

Fixed by cutting every returned row to the budget rather than only the elided one, which makes "no row exceeds the terminal" true for all input instead of for the common case. The width test now sweeps seven terminal widths against a narrow, a wide (`你`) and an astral (`😀`) filler; with the cut removed it fails at three columns with a measured width of 4.
